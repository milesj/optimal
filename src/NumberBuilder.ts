import Builder from './Builder';
import { DefaultValue } from './types';

function isNumber(value: unknown): value is number {
  return typeof value === 'number';
}

export default class NumberBuilder<T extends number = number> extends Builder<T> {
  constructor(defaultValue?: DefaultValue<T>) {
    super('number', defaultValue || (0 as T));
  }

  between(min: number, max: number, inclusive: boolean = false): this {
    if (__DEV__) {
      this.invariant(
        isNumber(min) && isNumber(max),
        'Between requires a minimum and maximum number.',
      );

      this.addCheck((path, value) => {
        this.invariant(
          isNumber(value) &&
            (inclusive ? value >= min && value <= max : value > min && value < max),
          `Number must be between ${min} and ${max}${inclusive ? ' inclusive' : ''}.`,
          path,
        );
      });
    }

    return this;
  }

  gt(min: number, inclusive: boolean = false): this {
    if (__DEV__) {
      this.invariant(isNumber(min), 'Greater-than requires a minimum number.');

      this.addCheck((path, value) => {
        if (inclusive) {
          this.invariant(
            isNumber(value) && value >= min,
            `Number must be greater than or equal to ${min}.`,
            path,
          );
        } else {
          this.invariant(
            isNumber(value) && value > min,
            `Number must be greater than ${min}.`,
            path,
          );
        }
      });
    }

    return this;
  }

  gte(min: number): this {
    return this.gt(min, true);
  }

  lt(max: number, inclusive: boolean = false): this {
    if (__DEV__) {
      this.invariant(isNumber(max), 'Less-than requires a maximum number.');

      this.addCheck((path, value) => {
        if (inclusive) {
          this.invariant(
            isNumber(value) && value <= max,
            `Number must be less than or equal to ${max}.`,
            path,
          );
        } else {
          this.invariant(isNumber(value) && value < max, `Number must be less than ${max}.`, path);
        }
      });
    }

    return this;
  }

  lte(max: number): this {
    return this.lt(max, true);
  }

  oneOf<U extends number>(list: U[]): NumberBuilder<U> {
    if (__DEV__) {
      this.invariant(
        Array.isArray(list) && list.length > 0 && list.every(item => isNumber(item)),
        'One of requires a non-empty array of numbers.',
      );

      this.addCheck((path, value) => {
        this.invariant(list.includes(value), `Number must be one of: ${list.join(', ')}`, path);
      });
    }

    return (this as unknown) as NumberBuilder<U>;
  }
}

export function number<T extends number = number>(defaultValue?: DefaultValue<number>) /* infer */ {
  return new NumberBuilder<T>(defaultValue as T);
}
