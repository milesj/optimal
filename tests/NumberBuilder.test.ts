import NumberBuilder, { number } from '../src/NumberBuilder';

describe('NumberBuilder', () => {
  let builder: NumberBuilder;

  beforeEach(() => {
    builder = number(123);
  });

  describe('constructor()', () => {
    it('sets default value', () => {
      expect(number(456).defaultValue).toBe(456);
    });
  });

  describe('runChecks()', () => {
    it('errors if a non-number value is used', () => {
      expect(() => {
        // @ts-ignore Testing wrong type
        number().runChecks('key', 'foo', {});
      }).toThrowErrorMatchingSnapshot();
    });
  });

  describe('between()', () => {
    it('errors if min is a not a number', () => {
      expect(() => {
        // @ts-ignore Testing wrong type
        builder.between('foo', 5);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if max is a not a number', () => {
      expect(() => {
        // @ts-ignore Testing wrong type
        builder.between(0, 'foo');
      }).toThrowErrorMatchingSnapshot();
    });

    it('adds a checker', () => {
      builder.between(0, 5);

      expect(builder.checks[1]).toEqual({
        callback: builder.checkBetween,
        args: [0, 5, false],
      });
    });
  });

  describe('checkBetween()', () => {
    it('errors if value is not a number', () => {
      expect(() => {
        // @ts-ignore Testing wrong type
        builder.checkBetween('key', 'foo', 0, 5);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if out of range', () => {
      expect(() => {
        builder.checkBetween('key', 10, 0, 5);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if out of range inclusive', () => {
      expect(() => {
        builder.checkBetween('key', 10, 0, 5, true);
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if in range', () => {
      expect(() => {
        builder.checkBetween('key', 3, 0, 5);
      }).not.toThrow('Invalid field "key". Number must be between 0 and 5.');
    });

    it('doesnt error if in range inclusive', () => {
      expect(() => {
        builder.checkBetween('key', 5, 0, 5, true);
      }).not.toThrow('Invalid field "key". Number must be between 0 and 5 inclusive.');
    });
  });

  describe('gt()', () => {
    it('errors if min is a not a number', () => {
      expect(() => {
        // @ts-ignore Testing wrong type
        builder.gt('foo');
      }).toThrowErrorMatchingSnapshot();
    });

    it('adds a checker', () => {
      builder.gt(5);

      expect(builder.checks[1]).toEqual({
        callback: builder.checkGreaterThan,
        args: [5, false],
      });
    });
  });

  describe('gte()', () => {
    it('errors if min is a not a number', () => {
      expect(() => {
        // @ts-ignore Testing wrong type
        builder.gte('foo');
      }).toThrowErrorMatchingSnapshot();
    });

    it('adds a checker', () => {
      builder.gte(5);

      expect(builder.checks[1]).toEqual({
        callback: builder.checkGreaterThan,
        args: [5, true],
      });
    });
  });

  describe('checkGreaterThan()', () => {
    it('errors if value is not a number', () => {
      expect(() => {
        // @ts-ignore Testing wrong type
        builder.checkGreaterThan('key', 'foo', 5);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if below minimum', () => {
      expect(() => {
        builder.checkGreaterThan('key', 3, 5);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if below minimum inclusive', () => {
      expect(() => {
        builder.checkGreaterThan('key', 3, 5, true);
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if above minimum', () => {
      expect(() => {
        builder.checkGreaterThan('key', 10, 5);
      }).not.toThrow('Invalid field "key". Number must be greater than 5.');
    });

    it('doesnt error if above minimum inclusive', () => {
      expect(() => {
        builder.checkGreaterThan('key', 5, 5, true);
      }).not.toThrow('Invalid field "key". Number must be greater than or equal to 5.');
    });
  });

  describe('lt()', () => {
    it('errors if max is a not a number', () => {
      expect(() => {
        // @ts-ignore Testing wrong type
        builder.lt('foo');
      }).toThrowErrorMatchingSnapshot();
    });

    it('adds a checker', () => {
      builder.lt(5);

      expect(builder.checks[1]).toEqual({
        callback: builder.checkLessThan,
        args: [5, false],
      });
    });
  });

  describe('lte()', () => {
    it('errors if max is a not a number', () => {
      expect(() => {
        // @ts-ignore Testing wrong type
        builder.lte('foo');
      }).toThrowErrorMatchingSnapshot();
    });

    it('adds a checker', () => {
      builder.lte(5);

      expect(builder.checks[1]).toEqual({
        callback: builder.checkLessThan,
        args: [5, true],
      });
    });
  });

  describe('checkLessThan()', () => {
    it('errors if value is not a number', () => {
      expect(() => {
        // @ts-ignore Testing wrong type
        builder.checkLessThan('key', 'foo', 5);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if above maximum', () => {
      expect(() => {
        builder.checkLessThan('key', 7, 5);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if above maximum inclusive', () => {
      expect(() => {
        builder.checkLessThan('key', 7, 5, true);
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if below maximum', () => {
      expect(() => {
        builder.checkLessThan('key', 3, 5);
      }).not.toThrow('Invalid field "key". Number must be less than 5.');
    });

    it('doesnt error if below maximum inclusive', () => {
      expect(() => {
        builder.checkLessThan('key', 5, 5, true);
      }).not.toThrow('Invalid field "key". Number must be less than or equal to 5.');
    });
  });

  describe('oneOf()', () => {
    it('errors if not an array', () => {
      expect(() => {
        // @ts-ignore Testing wrong type
        builder.oneOf(123);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if array is empty', () => {
      expect(() => {
        builder.oneOf([]);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if array contains a non-number', () => {
      expect(() => {
        // @ts-ignore Testing wrong type
        builder.oneOf(['foo', 123]);
      }).toThrowErrorMatchingSnapshot();
    });

    it('adds a checker', () => {
      builder.oneOf([123, 456, 789]);

      expect(builder.checks[1]).toEqual({
        callback: builder.checkOneOf,
        args: [[123, 456, 789]],
      });
    });
  });

  describe('checkOneOf()', () => {
    it('errors if value is not in the list', () => {
      expect(() => {
        builder.checkOneOf('key', 666, [123, 456, 789]);
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if value contains token', () => {
      expect(() => {
        builder.checkOneOf('key', 123, [123, 456, 789]);
      }).not.toThrow('Invalid field "key". Number must be one of: 123, 456, 789');
    });
  });

  describe('typeAlias()', () => {
    it('returns the type name', () => {
      expect(number().typeAlias()).toBe('number');
    });
  });
});
