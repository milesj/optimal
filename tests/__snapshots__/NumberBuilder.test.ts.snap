// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberBuilder between() errors if edge of range 1`] = `"Invalid field \\"key\\". Number must be between 0 and 5."`;

exports[`NumberBuilder between() errors if max is a not a number 1`] = `"Between requires a minimum and maximum number."`;

exports[`NumberBuilder between() errors if min is a not a number 1`] = `"Between requires a minimum and maximum number."`;

exports[`NumberBuilder between() errors if out of range 1`] = `"Invalid field \\"key\\". Number must be between 0 and 5."`;

exports[`NumberBuilder between() errors if value is not a number 1`] = `"Invalid field \\"key\\". Must be a number."`;

exports[`NumberBuilder between(): inclusive errors if out of range 1`] = `"Invalid field \\"key\\". Number must be between 0 and 5 inclusive."`;

exports[`NumberBuilder gt() errors if below minimum 1`] = `"Invalid field \\"key\\". Number must be greater than 5."`;

exports[`NumberBuilder gt() errors if edge 1`] = `"Invalid field \\"key\\". Number must be greater than 5."`;

exports[`NumberBuilder gt() errors if min is a not a number 1`] = `"Greater-than requires a minimum number."`;

exports[`NumberBuilder gte() errors if below minimum 1`] = `"Invalid field \\"key\\". Number must be greater than or equal to 5."`;

exports[`NumberBuilder gte() errors if min is a not a number 1`] = `"Greater-than requires a minimum number."`;

exports[`NumberBuilder lt() errors if above maximum 1`] = `"Invalid field \\"key\\". Number must be less than 5."`;

exports[`NumberBuilder lt() errors if edge 1`] = `"Invalid field \\"key\\". Number must be less than 5."`;

exports[`NumberBuilder lt() errors if max is a not a number 1`] = `"Less-than requires a maximum number."`;

exports[`NumberBuilder lte() errors if above maximum 1`] = `"Invalid field \\"key\\". Number must be less than or equal to 5."`;

exports[`NumberBuilder lte() errors if max is a not a number 1`] = `"Less-than requires a maximum number."`;

exports[`NumberBuilder oneOf() errors if array contains a non-number 1`] = `"One of requires a non-empty array of numbers."`;

exports[`NumberBuilder oneOf() errors if array is empty 1`] = `"One of requires a non-empty array of numbers."`;

exports[`NumberBuilder oneOf() errors if not an array 1`] = `"One of requires a non-empty array of numbers."`;

exports[`NumberBuilder oneOf() errors if value is not in the list 1`] = `"Invalid field \\"key\\". Number must be one of: 123, 456, 789"`;

exports[`NumberBuilder runChecks() errors if a non-number value is used 1`] = `"Invalid field \\"key\\". Must be a number."`;
